src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -858,49 +871,84 @@ tx_gateway_frontend::get_ongoing_tx(
+        if (!stm->is_actual_term(tx.etag)) {
+            // Intentnally empty body. Leaving it just to comment what's going
+            // on.
+            //
+            // We don't save ongoing state to the log so the only case when it's
+            // possible to observe old ongoing transaction is when the tx was
+            // started on this node. Then an another node became leader, a
+            // client didn't issue any new requests to that node, the current
+            // node became leader again and the client woke up. It's safe to
+            // continue it.
+        }
+        co_return tx;
+    } else if (tx.status == tm_transaction::tx_status::preparing) {
+        // a producer can see a transaction with the same pid and in a
+        // preparing state only if it attempted a commit, the commit
+        // failed and then the producer ignored it and tried to start
+        // another transaction.
+        //
+        // it violates the docs, the producer is expected to call abort
+        // https://kafka.apache.org/23/javadoc/org/apache/kafka/clients/producer/KafkaProducer.html
+        co_return tx_errc::timeout;
+    } else {
+        // A previous transaction has failed after its status has been
+        // decided, rolling it forward.
+        tx_errc ec;
+        if (tx.status == tm_transaction::tx_status::prepared) {
+            ec = co_await recommit_tm_tx(tx, timeout);
         } else if (tx.status == tm_transaction::tx_status::aborting) {
-            r = co_await reabort_tm_tx(tx, timeout);
+            ec = co_await reabort_tm_tx(tx, timeout);
         } else {
-            vassert(
-              tx.status == tm_transaction::tx_status::finished,
-              "unexpected tx status {}",
-              tx.status);
+            vassert(false, "unexpected tx status {}", tx.status);
         }
 
-        if (!r.has_value()) {
-            co_return r;
+        if (ec != tx_errc::none) {
+            co_return ec;
         }
 
-        auto tx = r.value();
-
-        auto changed_tx = stm->mark_tx_ongoing(tx.id);
-        if (!changed_tx.has_value()) {
-            co_return checked<cluster::tm_transaction, tx_errc>(
-              tx_errc::timeout);
+        if (!stm->is_actual_term(tx.etag)) {
+            // The tx has started on the previous term. Even though we rolled it
+            // forward there is a possibility that a previous leader did the
+            // same and already started the current transaction.
+            //
+            // Failing the current request. By the spec a client should abort on
+            // failure, but abort doesn't handle prepared and aborting statuses.
+            // So marking it as ready. We use previous term because aborting a
+            // tx in current ready state with current term means we abort a tx
+            // which wasn't started and it leads to an error.
+            [[maybe_unused]] auto _ = co_await stm->mark_tx_ready(
```

Comment:
you can do a plain `co_await stm->mark_tx_ready(` without assigning the return value to any variable?

Reply:
without it it's `error: ignoring return value of function declared with 'nodiscard' attribute`

Reply:
did you try the canonical `(void)co_await...` doesn't really matter, but for the future, `_` is still a variable introduced on the stack.

Reply:
I was sure I did but I was wrong, it works, thanks!